File: Day 2/Day 2.py
with open('data.txt', 'r') as r:
    lines = r.read().splitlines()

totalRed = 12
totalGreen = 13
totalBlue = 14

total = 0
totalProd = 0

def parseColor(game, color):
    if color not in game:
        return -1
    return int(game.split(" " + color)[0].split()[-1])

for i, line in enumerate(lines, start=1):
    
    games = line.split("; ")
    red = blue = green = -1
    maxRed = maxGreen = maxBlue = 0
    possible = True

    for game in games:
        red = parseColor(game, "red")
        green = parseColor(game, "green")
        blue = parseColor(game, "blue")

        if red > totalRed or green > totalGreen or blue > totalBlue:
            possible = False
        
        maxRed = max(maxRed, red)
        maxGreen = max(maxGreen, green)
        maxBlue = max(maxBlue, blue)
    
    totalProd += maxRed * maxGreen * maxBlue
    if possible:
        total += i

print("Part 1:", total)
print("Part 2:", totalProd)

File: Day 2/Day 2.go
package main

import (
	"bufio"
	"fmt"
	"math"
	"os"
	"strconv"
	"strings"
)

func parseColor(game string, color string) int {
	if !strings.Contains(game, color) {
		return -1
	}
	colorStr := strings.Split(strings.Split(game, " "+color)[0], " ")
	num, _ := strconv.Atoi(colorStr[len(colorStr)-1])
	return num
}

func main() {
	data, err := os.Open("data.txt")

	if err != nil {
		fmt.Println("File reading error", err)
		return
	}

	defer data.Close()

	scanner := bufio.NewScanner(data)
	scanner.Split(bufio.ScanLines)

	var curGame int = 1
	var totalRed int = 12
	var totalGreen int = 13
	var totalBlue int = 14
	var total int = 0
	var totalProd int = 0

	for scanner.Scan() {

		var line string = scanner.Text()
		games := strings.Split(line, "; ")
		possible := true
		var maxRed, maxGreen, maxBlue int = 0, 0, 0
		for _, game := range games {
			red := parseColor(game, "red")
			blue := parseColor(game, "blue")
			green := parseColor(game, "green")
			if (red > totalRed) || (blue > totalBlue) || (green > totalGreen) {
				possible = false
			}

			maxRed = int(math.Max(float64(maxRed), float64(red)))
			maxGreen = int(math.Max(float64(maxGreen), float64(green)))
			maxBlue = int(math.Max(float64(maxBlue), float64(blue)))
		}

		if possible {
			total += curGame
		}

		totalProd += maxRed * maxGreen * maxBlue
		curGame++
	}
	fmt.Println("Part 1:", total)
	fmt.Println("Part 2:", totalProd)
}
